feat: Create Record and Replay ADR (#863)
edgexfoundry/edgex-docs: docs_src/design/adr/application/0025-Record-and-Replay.md

# Record and Replay ADR

## Submitters
- Lenny Goodell (Intel)

## Change Log
- [Pending](https://github.com/edgexfoundry/edgex-docs/pull/863) (2022-09-20)

## Referenced Use Case(s)
- [Record and Replay UCR](https://docs.edgexfoundry.org/2.3/design/ucr/Record-and-Replay/)

## Context

> This design involves creating a new Application Service that is responsible for the requirements in the above referenced UCR. This new Application Service will go above and beyond that of a normal Application Service,  thus it is architecturally significant.

Reviewer: and above and beyond the capabilities of the virtual device service

Author: Not sure why you are mentioning virtual device service here. This section is about justifying `architecturally significant`.

> This design involves creating a new Application Service that is responsible for the requirements in the above referenced UCR. This new Application Service will go above and beyond that of a normal Application Service,  thus it is architecturally significant.

Reviewer 2: > This new Application Service will go above and beyond that of a normal Application Service,  thus it is architecturally significant.

I don't think creating a complex App Service justifies needing an ADR. I'd just remove this and justify the need for an ADR (as mentioned by Lenny in the TSC call) by saying this is a first-party service and need a formal design review.

Author: Updated

## Proposed Design

> A new Application Service will be created with a RESTful API to handle the Record, Replay, Export and Import capabilities. The service will not create or start a Functions Pipeline on start-up as normally done in Application Services. It will wait until the Record request has been received. Once the recording is complete the Functions Pipeline will be stopped. 

Reviewer 2: Why should this be a service and have a RESTful API to control the operations? Why not make this an application which takes CLI sub-commands (record, replay, import, export) and arguments or a configuration file and can be gracefully shutdown on request? **Why can't this be simply added to the EdgeX CLI?** The UCR did not mention the need for a service or list any requirement for network-based / remote control of the operations.

This could be justified if there was a need for parallel recording sessions.

A application is superior because it can also offer logging to the user to inform them about the progress.

Author: See my response to your same general comment.

Reviewer 2: Your [comment](https://github.com/edgexfoundry/edgex-docs/pull/863#issuecomment-1261328653) does not answer any of my questions.

Author: Not sure what you are looking for here.  Correct the UCR doesn't have a requirement of service via cli, that is a design choice.

> This could be justified if there was a need for parallel recording sessions.

This could be a future enhancement

> A application is superior because it can also offer logging to the user to inform them about the progress.

Yes, this is correct. Expect implementation to have logging to show progress.

Reviewer 2: CLI vs networking API isn't a design choice. CLI is a user-facing interface of an application, while a networking API for machines to communicate with a web service. They serve very different purposes and have different use cases.

If we are offering a web service, the progress info is only useful if it is available also over the networking API; for example, an API call to query the progress.

Author: > the progress info is only useful if it is available also over the networking API

Yes, I agree that status endpoint would be very useful. I have added them.

Author: > CLI vs networking API isn't a design choice.

How can it not be a design choice. IMO it is a major design choice. UCR doesn't dictate it be designed using CLI or REST API.

Author: The design could call for a RESTful service or a CLI, that address the requirements, thus it is a design decision.

Author: > Why should this be a service

This has to be a **service** as it must connect  to the MessageBus and consume Events over a long period of time (just like other App Services). A CLI would have to duplicate this long run service in the background which is not very CLI'ish.

Reviewer 2: > > Why should this be a service
> 
> This has to be a **service** as it must connect to the MessageBus and consume Events over a long period of time (just like other App Services). A CLI would have to duplicate this long run service in the background which is not very CLI'ish.

Okay, at last! That is a compelling reason which should be documented in the design.

Author: Added to `Proposed Design` section and updated CLI item in `Considerations` section.

> A new Application Service will be created with a RESTful API to handle the Record, Replay, Export and Import capabilities. The service will not create or start a Functions Pipeline on start-up as normally done in Application Services. It will wait until the Record request has been received. Once the recording is complete the Functions Pipeline will be stopped. There will be a very rudimentary HTML page to serve as UI to access this service RESTful API.

Reviewer: Concern with bringing in packages to provide any type of display and not sure it is needed if you have the APIs (and postman).  But I defer to the community on this one.

> A new Application Service will be created with a RESTful API to handle the Record, Replay, Export and Import capabilities. The service will not create or start a Functions Pipeline on start-up as normally done in Application Services. It will wait until the Record request has been received. Once the recording is complete the Functions Pipeline will be stopped. There will be a very rudimentary HTML page to serve as UI to access this service RESTful API.

Reviewer 3: Thinking about the "rudimentary HTML page" that our product has for a similar recorder, we do have CGI scripts (with user role checking) behind the simple push buttons, so I suppose it can't be too rudimentary. I'd support dropping this, too, then.

> A new Application Service will be created with a RESTful API to handle the Record, Replay, Export and Import capabilities. The service will not create or start a Functions Pipeline on start-up as normally done in Application Services. It will wait until the Record request has been received. Once the recording is complete the Functions Pipeline will be stopped. There will be a very rudimentary HTML page to serve as UI to access this service RESTful API.

Author: I am good with this!!!

!!! note
    Application Services do not receive data when the Functions Pipelines are stopped.

### Record Endpoint

> This `POST` API will start recording data as specified in the [request Data Transfer Object (DTO)](#record-request-dto) defined below. The request handler will validate the DTO and then create a new Functions Pipeline and [Start the Functions Pipeline](https://docs.edgexfoundry.org/2.3/microservices/application/ApplicationServiceAPI/#makeitrun) to process incoming data. 

Author: The recording auto stops. Are you thinking we need a Cancel, which was not in the approved UCR requirements. Could be a later enhancement once we get folks actually using it.

> This `POST` API will start recording data as specified in the [request Data Transfer Object (DTO)](#record-request-dto) defined below. The request handler will validate the DTO and then create a new Functions Pipeline and [Start the Functions Pipeline](https://docs.edgexfoundry.org/2.3/microservices/application/ApplicationServiceAPI/#makeitrun) to process incoming data. 

Reviewer: Ok with this being future consideration - but yes, how do you stop or cancel the recording when perhaps you don't want it to go to the full length?

> This `POST` API will start recording data as specified in the [request Data Transfer Object (DTO)](#record-request-dto) defined below. The request handler will validate the DTO and then create a new Functions Pipeline and [Start the Functions Pipeline](https://docs.edgexfoundry.org/2.3/microservices/application/ApplicationServiceAPI/#makeitrun) to process incoming data. 

Author: Stopping the restarting the service would do it. Larger hammer for now... ;-)

The Functions Pipeline will contain the following pipeline functions in the following order

- [Filter functions](https://docs.edgexfoundry.org/2.3/microservices/application/BuiltIn/#filtering) if filtering is needed, configured based on the DTO parameters.
- [Batching pipeline function](https://docs.edgexfoundry.org/2.3/microservices/application/BuiltIn/#batching) configured based on the DTO parameters. This will be used to control the record duration/count. 

> - New pipeline function written to process the batched data once the batching threshold has been exceeded. This function will simply send the recorded data to a channel to be processed. 

Reviewer 3: "to a channel" - I don't see what this means; is it apparent to everyone else?

Author: It is a Go thing for async communication. There is an asyc function reading message from the specified channel.
How about this: `send the recorded data to aysnc function to be processed`

Reviewer 3: That sounds good - makes sense - thanks!

> The async  function receiving the data will first [stop the Functions Pipeline](https://docs.edgexfoundry.org/2.3/microservices/application/ApplicationServiceAPI/#makeitstop) and then save the data for later replay and/or export. It will also determine the list of unique Device Profile and Device Names from the data and store them along side the recorded data. Since app services can receive Events out of order per their timestamps, the saved Event data must be sorted by the Event timestamps. All data will saved in in-memory storage. 

Reviewer 2: > All data will saved in in-memory storage. 

This again makes it harder to justify having this component as a service with a networking API. Maybe I misunderstood, but if all events are buffered in memory, it is better to allow execution only on device to an administrator that runs this process in foreground with knowledge of available system resources.

Author: > Maybe I misunderstood, but if all events are buffered in memory, it is better to allow execution only on device to an administrator that runs this process in foreground with knowledge of available system resources.

This is a developer tool, not something that is intended to run on a production system.

Reviewer 2: A developer tool which runs as a service needs to be implemented to function reliably in background. 

> Since app services can receive Events out of order per their timestamps, the saved Event data must be sorted by the Event timestamps. All data will saved in in-memory storage.

Is the data stored in memory to simplify sorting? Why not use a key-value such as [LevelDB](https://github.com/syndtr/goleveldb) which persists on disk and inserts in sorted order?

Reviewer 2: A developer tool which runs as a service needs to be implemented to function reliably in background. 

> Since app services can receive Events out of order per their timestamps, the saved Event data must be sorted by the Event timestamps. All data will saved in in-memory storage.

Is the data stored in memory to simplify sorting? Why not use a key-value such as [LevelDB](https://github.com/syndtr/goleveldb) which persists on disk and inserts in sorted order?

Author: > Is the data stored in memory to simplify sorting?

No, it is to keep the initial implementation simple (crawl phase). I am not oppose to a more complex storage as a follow on enhancement once we determine it is justified by usage and feedback.

Author: If removing specifying the type of storage appease you here then I am fine with leaving it to an implementation detail.

Reviewer 3: I found it clarifying that the ADR said it stored the capture in memory; otherwise, it looked like the design missed something where it would have needed a means to specify a filename to save it as.

Reviewer 2: If you truly believe in-memory recording is good design and worth becoming part of the EdgeX reference architecture for record and reply, then please leave it.

Reviewer 4: in-memory recording could be an easy starting of this service, and the persistence layer can be extended in the future.
however, according to my experience, if the event frequency is high, for example, 100 auto events with 100ms interval, the database access could be an serious bottleneck, even using the memory db like Redis or [ql(an embedded SQL database engine)](https://godoc.org/github.com/cznic/ql).

We did a big change for device-virtual to improve the performance.
https://github.com/edgexfoundry/device-virtual-go/issues/260

!!! note

>     Starting a new recording will overwrite any previous recorded data.

Reviewer 2: This implies that we don't wanna support parallel sessions, as opposed to what someone may expect from a web service. I would either allow parallel recordings or throw an error if a subsequent request is sent, forcing the user to DELETE before POSTing another. On the other hand, a PUT operation can be used to "replace" an existing recording.

Author: This not about multiple recording at once, it is about there only being one data set that will be overwritten.

> This implies that we don't wanna support parallel sessions

Correct, this first implementation is a crawl phase. Once we get usage and feedback we can add support for support parallel sessions.

> throw an error if a subsequent request is sent, forcing the user to DELETE before POSTing another

Yes, an error will be returned if a recording is already in progress. Will add wording for that (same for Replat) . The DELETE APIs to cancel a recording or replay have already been added.

#### Record Request DTO

##### Duration

Time duration in which to record data. Required if **Event Limit** is not specified.

##### Event Limit 

Maximum number `Events` to record. Required if **Duration** is not specified

##### Include Device Profile Names 

Optional list of Device Profile Names to filter for

##### Include Device Names

Optional list of Device Names to filter for

##### Exclude Device Profile Names

Optional list of Device Profile Names to filter out

##### Exclude Device Names

Optional list of Device Names to filter out

### Replay endpoint

> This `POST` API will start replaying the recorded data as specified in the [request Data Transfer Object (DTO)](#replay-request-dto) defined below. The request handler will validate the DTO and that the appropriate Device Profiles and Devices from the data exist. It will then start an async Go function to handle the replay so the request doesn't timeout on long replays. 

Reviewer: do we anticipate a replay to the end of the recording or could someone ask to stop the replay mid-stream?  If so, would assume we need APIs for that?

Author: Replay will do the complete recording. Canceling the replay is not a requirement in the approved UCR. As with above this could be a future enhancement base on user requests.

> The replay async Go function will use the [Background Publishing](https://docs.edgexfoundry.org/2.3/microservices/application/AdvancedTopics/#background-publishing) capability to send the recorded Events to the EdgeX MessageBus using the same publish topic scheme used by Device Services, which is `edgex/events/device/<device-profile-name>/<device-name>/<source-name>`. The App SDK has the  [Publish Topic Placeholders](https://docs.edgexfoundry.org/2.3/microservices/application/Triggers/#publish-topic-placeholders) capability built-in to facilitate this. The data for these topics is available from the Event DTO. The timestamps in the Events and Readings published will be set to the current date/time. This requires a copy be made of the Event/Readings as they are publish in order to not corrupt the original data.

Reviewer 5: ```suggestion
The replay async Go function will use the [Background Publishing](https://docs.edgexfoundry.org/2.3/microservices/application/AdvancedTopics/#background-publishing) capability to send the recorded Events to the EdgeX MessageBus using the same publish topic scheme used by Device Services, which is `edgex/events/device/<device-profile-name>/<device-name>/<source-name>`. The App SDK has the  [Publish Topic Placeholders](https://docs.edgexfoundry.org/2.3/microservices/application/Triggers/#publish-topic-placeholders) capability built-in to facilitate this. The data for these topics is available from the Event DTO. The timestamps in the Events and Readings published will be set to the current date/time. This requires a copy be made of the Event/Readings as they are published in order to not corrupt the original data.
```

> Once the first event is published the replay function will calculate the wait time to use before sending the next Event from the recorded data. This will be based on the time difference from the original timestamp of the previous event published and the timestamp of the next event multiplied by the inverse of the `Replay Rate` specified in the request DTO. 

Reviewer: Really like this!  I assume it is our best effort at hitting the replay rate as desired.  Given we are not deterministic and running on POSIX like threads or something, we can only attempt to replicate the rates.   Something we should mention in the docs that this is not completely deterministic and depends on the machine and what it is doing.

!!! example - "Examples - Replay Rate wait time calculation"
    Delta time between original Events is 800ms<br/>
    Replay rate is 2.0 (100% faster) making wait time 400ms (800ms * (1 / 2.0))<br/>
    Replay rate is 0.5 (100% slower) making wait time 1600ms (800ms * (1 / 0.5))

The replay function will repeat publishing the recorded data per the `Repeat Count` in from the DTO. 

#### Replay Request DTO

##### Replay Rate

Required rate at which to replay the data compared to the rate the data was recorded. Float value greater than 0 where 1 is the same rate, less than 1 is slower rate and greater than 1 is faster rate than the rate the data was recorded. 

##### Repeat Count

> Optional count of number of times to repeat the replay. Default is 1 if not specified or value is 0.

Reviewer 2: ```suggestion
Optional count of number of times to repeat the replay. Defaults to 1 if not specified or set to 0.
```

### Download endpoint (Export)

> This `GET` API will request that the previously recorded data be exported as a file download. It will accept an optional query parameter to specify compression (NONE, ZIP or GZIP). An error is returned if no data has been recorded or invalid compression type requested.

Reviewer 2: In HTTP requests, the compression is usually requested using the [Accept-Encoding](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Encoding) header.

Author: > Having a web service adds several complications for no obvious gain.

I disagree. Having a service makes it more accessible. If you prefer the CLI experience you can use curl commands or as I prefer POSTMan or the UI can have a tab to control the service.

Author: [Accept-Encoding](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Encoding) tells the client what the service supports, not what the client is requesting.
IMO, the query parameter is appropriate.

Reviewer 2: No, that is not correct.
> The Accept-Encoding request HTTP header indicates the content encoding (usually a compression algorithm) that the client can understand. 

> IMO, the query parameter is appropriate.
If you want to not follow the standard practices and make your own, yes it is appropriate.

Author: It is more complicated than that. The server uses [content negotiation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation) to select one of the proposals and informs the client of that choice with the [Content-Encoding](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding) response header. 

Yes, I am fine with "none standard" if that is what you feel it is. ;-)

> This `GET` API will request the previously recorded data as a file download. It will accept an optional query parameter to specify composition (NONE, ZIP or GZIP). An error is returned if no data has been recorded or invalid compression type requested.

Reviewer 3: I had to read it three times to get the right sense of this, so suggesting a revision for clarity.

> This `GET` API will request the previously recorded data as a file download. It will accept an optional query parameter to specify composition (NONE, ZIP or GZIP). An error is returned if no data has been recorded or invalid compression type requested.

Author: Well, `composition` should be `compression type` ;-) 
Does that help clarity?

> This `GET` API will request the previously recorded data as a file download. It will accept an optional query parameter to specify composition (NONE, ZIP or GZIP). An error is returned if no data has been recorded or invalid compression type requested.

Reviewer 3: I missed "composition" myself; yes, that helps. Still recommend improving the first part, so the direction is clear. "export" and "GET" are not a natural pairing in my mental mapping, but your choice of GET is correct here, upon reflection.

> This `GET` API will request the previously recorded data as a file download. It will accept an optional query parameter to specify composition (NONE, ZIP or GZIP). An error is returned if no data has been recorded or invalid compression type requested.

Author: Ahh Yep, goo point. Will rename the Export/Import endpoint to download/upload for better clarity for users.

> The file content will be the [Recorded Data DTO](#recorded-data-dto) as define below. The request handler will build the DTO described below by extracting the recorded `Events` from in-memory storage, pulling the referenced `Device Profiles` and `Devices` from Core Metadata using the names from in-memory storage. The file extension used will be `.json`, `.zip` or `.gzip` depending on the compression selected.

Reviewer 2: More of an implementation concern: Data collection in memory, compression, and serialization together will be a recipe for memory exhaustion. I suggest making use of the secondary storage and data streaming techniques for some if not all the operations. The data can be streaming to an HTTP client too; see chunked transfer encoding: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding.

Author: > I suggest making use of the secondary storage
Secondary storage can be an enhancement (walk phase) once we get some usage and feedback on the need.

Author: > I suggest making use of the secondary storage and data streaming techniques for some if not all the operations. The data can be streaming to an HTTP client too; see chunked transfer encoding: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding.

These are implementation details which we'll consider during the implementation.

#### Recorded Data DTO

##### Events

List of `Events` (with `Readings`) that were recorded

##### Device Profiles

List of `Device Profiles` (complete profiles) that are referenced in the recorded `Events` 

##### Devices

List of `Device defintions` that are referenced in the recorded `Events` 

### Upload endpoint (Import)

This `POST` API will upload previously exported recorded data file. It will accept an optional Boolean query parameter to specify to **not** overwrite existing Device Profiles and/or Devices if they already exist. Default is to overwrite existing with those captured with the recorded data.

> The request handler will receive the file as a [Recorded Data DTO](#recorded-data-dto) described above and detect if it is compressed and un-compress the contents if needed before un-marshaling the JSON into the DTO. The compression will be determined based on the file extension. The `Event` data from the DTO will then be saved to the in-memory storage along with the Device Profile and Device Names. The `Device Profiles` and `Devices` will be pushed to Core Metadata if they don't exist or if overwrite is enabled.  

Reviewer 2: > The compression will be determined based on the file extension.

Is there a file extension in the HTTP request? I believe the compression type should be set as the [Content-Encoding](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding) header.

Author: > Is there a file extension in the HTTP request?

When the file is uploaded, the request receives the file name along with the contents, but I do like the use of the [Content-Encoding](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding) header.

Reviewer 2: The extension is a metadata in Windows file system, not in Unix-based ones. In unix-like system, the file can contain anything and have any name. The Content-Encoding is the standard way of telling the server about the encoding format.

Author: I have already update the ADR  to describe use of the `Content-

!!! note
    Import will overwrite any previous recorded data.

## Considerations

> - Only one recorded data set is available for replay.

Reviewer 3: ```suggestion
- Only one recorded data set may be replayed at a time.
```

Author: @TomBrennan-Eaton , `may be replayed at a time` infers there may be multiple recorded. Keeping this at crawl phase supporting only a single recording.

Reviewer 3: To me, "available" would mean recordings I had exported. I think I see what you mean, maybe "Only one recorded data set may be imported and held in memory for replay."

Author: updated

- The whole data set is replayed. Can not specify to replay data for specific Devices within the larger data set.

> - Wait times simulating rate of Events published will not be perfect since dependent on non-Realtime OS.

Reviewer: answers my comment from above!  :)
